### PR TITLE
Media & Text Block: Fix deprecation with `isStackOnMobile` default value changed

### LINKED ISSUE #29109: Media & Text Block: Invalid content for blocks made before WP 5.4
## Description
When opening the editor for a post that contains a Media & Text block that was made in WP core 5.3 or earlier and was left unchanged since, the migration to the new version of the block does not work anymore in WP 5.6. The editor shows the message 'This block contains unexpected or invalid content.' with the 'Attempt block recovery' button.

## Step-by-step reproduction instructions
1. In WP 5.3 create a post and add a Media & Text block. Don't change any settings - this means the 'Stack on mobile' is false, which was the default at the time.
2. Save the post and close the editor screen.
3. Update to WP 5.6.
4. Open the post again.

## Expected behaviour
The Media & Text block code gets updated to the new version (visible in the console). This adds the missing `isStackedOnMobile` attribute, for example:
`<!-- wp:media-text {"mediaId":6,"mediaType":"image","isStackedOnMobile":false,"className":"alignwide"} -->`

## Actual behaviour
The Media & Text block is shown as invalid.
The console explains why, for example:
```
Block validation: Expected attribute `class` of value `wp-block-media-text is-stacked-on-mobile alignwide`, saw `wp-block-media-text alignwide`.

Block validation: Block validation failed for `core/media-text` ( 
Object { name: "core/media-text", icon: {…}, keywords: (2) […], providesContext: {}, usesContext: [], styles: [], category: "media", title: "Media & Text", description: "Set media and words side-by-side for a richer layout.", example: {…}, … }
 ).
Content generated by `save` function:
<div class="wp-block-media-text is-stacked-on-mobile alignwide"><figure class="wp-block-media-text__media"><img src="http://dev.wp-20210218.org/wp-content/uploads/2021/02/testing-gh.jpg" alt=""/></figure><div class="wp-block-media-text__content"></div></div>
Content retrieved from post body:
<div class="wp-block-media-text alignwide"><figure class="wp-block-media-text__media"><img src="http://dev.wp-20210218.org/wp-content/uploads/2021/02/testing-gh.jpg" alt="" class="wp-image-6"/></figure><div class="wp-block-media-text__content"></div></div>
```

## WordPress information
- WordPress version: 5.6
- Gutenberg version: Not installed
- Are all plugins except Gutenberg deactivated? Yes
- Are you using a default theme (e.g. Twenty Twenty-One)? Yes

## Device information
- Device: Desktop
- Operating system: Windows 10
- Browser: Firefox 85.0.2

### PR DESCRIPTION
**Pull request by t-hamano**
Fixes #29109

## What?

This PR fixes a broken deprecation related to the default value of `isStackedOnMoobile` changing from `false` to `true`.

## Why?

In #14364, the default value of the `isStackedOnMobile` attribute has been updated from `false` to `true`. At the same time, deprecations were also made and [the based attributes were defined as an object named `baseAttributes`](https://github.com/WordPress/gutenberg/pull/14364/files#diff-1f1d3f8d14ce5f982caa2de0c762395d1a87a02e8db1ebf11fcbe4b8e2d07331R22-R64). The `baseAttributes` variable is the base definition for the future deprecation, so I think this object must not be updated. 

However, in #26041 the default value of `isStackOnMobile` for `baseAttributes` was changed to `true`. This means that blocks from `v0` to `v2`, where the default value of `isStackOnMobile` is `false`, are no longer migrated.

## How?

I have reverted the definition of `baseAttributes` back to the original definition and renamed it to `v0Atributes` to prevent the same problem from occurring in the future. At the same time, I added a property to override `baseSAttributes` in v3 and later, where the default value of `isStackedOnMobile` is expected to be `true`.

## Testing Instructions

In the code editor, insert a version of the markup in which the default value of `isStackedOnMobile` is `false`:

```html
<!-- wp:media-text -->
<div class="wp-block-media-text alignwide"><figure class="wp-block-media-text__media"></figure><div class="wp-block-media-text__content"><!-- wp:paragraph {"placeholder":"Content…","fontSize":"large"} -->
<p class="has-large-font-size"></p>
<!-- /wp:paragraph --></div></div>
<!-- /wp:media-text -->
```

Verify that the markup updates as follows when the focus is removed from the code editor:

```html
<!-- wp:media-text {"align":"wide","isStackedOnMobile":false,"className":"alignwide"} -->
<div class="wp-block-media-text alignwide"><figure class="wp-block-media-text__media"></figure><div class="wp-block-media-text__content"><!-- wp:paragraph {"placeholder":"Content…","fontSize":"large"} -->
<p class="has-large-font-size"></p>
<!-- /wp:paragraph --></div></div>
<!-- /wp:media-text -->
```

### CODE REVIEW

**Review by aaronrobertshaw — COMMENTED** [review]
Nice work deciphering the history here and fixing the deprecations @t-hamano 👍 

The changes make sense to me and the PR branch tests as advertised.

I think before this lands, we should add a deprecation fixture for the media and text block that covers this scenario.

**Comment by t-hamano** [comment]
Thanks for the PR, @aaronrobertshaw!

> I think before this lands, we should add a deprecation fixture for the media and text block that covers this scenario.

I ran `fixtures:regenerate` and no diff occurred, perhaps the fixtures generated by #14364 are still valid?

**Comment by aaronrobertshaw** [comment]
I ran the fixture tests on trunk and got no failures. To me, that indicates that we have gap there that we should have covered by a new fixture or there wouldn't be anything to fix.

Taking a quick glance at https://github.com/WordPress/gutenberg/pull/14364, it appears that all the fixtures were updated to the new default state. None were left in the deprecated state where `isStackOnMobile` defaulted to `false` and no new deprecation was added to cover it. 

I still believe we need it here in this PR as a result.

**Comment by t-hamano** [comment]
> Taking a quick glance at #14364, it appears that all the fixtures were updated to the new default state. None were left in the deprecated state where isStackOnMobile defaulted to false and no new deprecation was added to cover it.

I see, I may have understood.

Do you mean that in #14364, we should have just added a new fixture file to test deprecation about `isStackOnMobile` without updating the existing fixture file?

**Comment by aaronrobertshaw** [comment]
> Do you mean that in https://github.com/WordPress/gutenberg/pull/14364, we should have just added a new fixture file to test deprecation about isStackOnMobile without updating the existing fixture file?

The existing fixtures needed updating so that they only covered their respective test cases. However by updating every fixture to reflect the new isStackOnMobile default and not adding a new fixture to cover the scenario with the old default, we didn't have anything to catch the issue in our deprecations.

Because that fixture wasn't added, we need to add it now.

So in other words:

- Yes, [#14364](https://github.com/WordPress/gutenberg/pull/14364) should have added a new fixture
- Yes, [#14364](https://github.com/WordPress/gutenberg/pull/14364) was correct to update the other existing fixtures
- Yes, we need to add the missing fixture covering the old default value

I hope that clears things up some 🙂

**Comment by t-hamano** [comment]
Thank you for the detailed explanation, @aaronrobertshaw!

I have added a new scenario for the block where the default value of `isStackOnMobile` is `false`.

- When "Stack on mobile" is enabled in the old block, the comment delimiter stores `"isStackOnMobile": true`.
- When the HTML is serialized, the default value of `isStackOnMobile` is `true` and `"isStackOnMobile": true` is removed from the comment delimiter.
- In the latest block, `alignwide` is not the default, so `"align":"wide"` is added to the comment delimiter.

Does this scenario make sense?

File: test/integration/fixtures/blocks/core__media-text__deprecated-v2.html [hunk 1]
```diff
@@ -0,0 +1,12 @@
+<!-- wp:media-text {"mediaType":"image", "isStackOnMobile": true} -->
+<div class="wp-block-media-text alignwide is-stacked-on-mobile">
```
Comment (aaronrobertshaw) [inline test/integration/fixtures/blocks/core__media-text__deprecated-v2.html]:
```suggestion
<!-- wp:media-text {"mediaType":"image"} -->
<div class="wp-block-media-text">
```

This fixture isn't covering the old v2 version. That version defaulted `isStackedOnMobile` to `false`. We should remove the explicit setting of the attribute in the block delimiter and the incorrect CSS class. 

With those removed, the block's markup will match the results of the v2 deprecation's save function using the old v2 attributes definition (`isStackedOnMobile: false`). After regenerating the fixtures, the final serialized HTML will match the current version i.e. `isStackedOnMobile` will be set to `false` explicitly within the block delimiter.

**Review by aaronrobertshaw — COMMENTED** [review]
Sorry for the delay in getting back to this @t-hamano 🙏 

I don't think we have the v2 deprecation fixture correct just yet. It needs to cover the old v2 version that was deprecated. 

That is, it should cover a block that was saved without `isStackedOnMobile ` being explicitly set so it would use the old `false` default value. This fixture then ensures the old v2 version without `isStackedOnMobile ` set, and therefore no `is-stacked-on-mobile` class in its markup, gets migrated to include the `isStackedOnMobile: false` in the block delimiter.

I've left an inline comment suggesting the above tweak. After applying it the other fixture files will need to be regenerated: `npm run fixtures:regenerate`

**Comment by t-hamano** [comment]
Thanks for the detailed explanation, @aaronrobertshaw! I have updated the fixture file regarding the v2 deprecation.

**Review by aaronrobertshaw — APPROVED** [review]
Thanks for the hard work wrestling with these deprecations @t-hamano, I appreciate it 👍 

The fixture tests are all passing for me. The changes look good.

I think we're ready to land this one. 

🚢

**Comment by t-hamano** [comment]
Thanks for the review, @aaronrobertshaw!

**Pull request merged**